examples/stm32f4_discovery/can_m2515/main.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2013, Kevin Läufer
+ * Copyright (c) 2013-2017, Niklas Hauser
+ * Copyright (c) 2016, Raphael Lehmann
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#include <modm/board.hpp>
+#include <modm/driver/can/mcp2515.hpp>
+#include <modm/processing/protothread.hpp>
+
+// If you use a different SPI instance, you may have to also choose different
+// GPIOs to connect to.
+using Int = GpioOutputB11;
+using Cs = GpioOutputB12;
+using Sck = GpioOutputB13;
+using Mosi = GpioOutputB15;
+using Miso = GpioInputB14;
+using SpiMaster = SpiMaster2;
+// Note that you can also use a bit-banged SPI driver as a drop-in replacement
+// using SpiMaster = BitBangSpiMaster<Sck, Mosi, Miso>;
+
+// Default filters to receive any extended CAN frame
+FLASH_STORAGE(uint8_t canFilter[]) = {
+	MCP2515_FILTER_EXTENDED(0),  // Filter 0
+	MCP2515_FILTER_EXTENDED(0),  // Filter 1
+
+	MCP2515_FILTER_EXTENDED(0),  // Filter 2
+	MCP2515_FILTER_EXTENDED(0),  // Filter 3
+	MCP2515_FILTER_EXTENDED(0),  // Filter 4
+	MCP2515_FILTER_EXTENDED(0),  // Filter 5
+
+	MCP2515_FILTER_EXTENDED(0),  // Mask 0
+	MCP2515_FILTER_EXTENDED(0),  // Mask 1
+};
+
+modm::Mcp2515<SpiMaster, Cs, Int> mcp2515;
+
+class MyTask : modm::pt::Protothread
+{
+public:
+	MyTask() : message_{0x123456}
+	{
+		mcp2515.initialize<8_MHz, 125_kbps>();
+		mcp2515.setFilter(modm::accessor::asFlash(canFilter));
```

Comment:
Doing stuff in the contructors is not recommended and won't work here because `MyTask` is constructed before the SPI is initialized (in `main()`).

```suggestion
	void
	initialize()
	{
		mcp2515.initialize<8_MHz, 125_kbps>();
		mcp2515.setFilter(modm::accessor::asFlash(canFilter));
```

Maybe it would also be a good idea to move the (blocking) spi stuff from `initialize()` to the first call of `update()`!?

Reply:
well i will  find a better place for it :)

Reply:
At the begin of the protothread (in `run()` after `PT_BEGIN()`)? I don't know why I didn't suggest this earlier... 🤦🏽

Reply:
Hmm I really don't want the user code to do that .. shouldn't this be somewhere inside the library ?

Reply:
i followed your suggestion now

Reply:
I'm ok with not fixing that in this PR.